### DISCUSSION TITLE
How to implement custom attributes

### DISCUSSION THREAD

**Discussion by Apocryphon-X** [vlang/v#15132] [General]
Is there a way to use/create custom attributes? There doesn't seem to be anything about this in the docs, so I'm asking.
(I want to try something similar to the vweb endpoints)
```v
['/hello/:user'] // <--------
fn (mut app App) hello_user(user string) vweb.Result {
	return app.text('Hello $user')
}
```

**Comment by archandha** [comment]
I didn't find any documentation neither. But you might look into https://github.com/vlang/v/blob/master/vlib/vweb/vweb.v in the functions `run_at<T>` and `handle_conn<T>` which both do this trick. 
In the current version of the file, there is an example in `run_at<T>` lines 397..406.

**Reply by Apocryphon-X** [reply]
Thanks! That reference helped me a lot. I think that's the only example available in the repository

**Comment by David-Worboys** [comment]
Could you use a map?

**Reply by Apocryphon-X** [reply]
I used this idea to make a very tiny module to simplify compile time reflection

**Comment by Apocryphon-X** [comment]
After looking at the `vweb` source code, I wrote a custom attribute example for any people with the same question in the future. I hope this helps anyone!
```v
import reflection { methods_of_type }

struct Dummy {}

[example: 'Lorem Ipsum']
fn (s Dummy) foo() {
	// ...
}

fn main() {
	dummy_methods := methods_of_type<Dummy>()
	assert 'foo' in dummy_methods

	dump(dummy_methods['foo'].attrs)
	dump(dummy_methods['foo'])
}
```
Output:
```v
[main.v:15] dummy_methods['foo'].attrs: ['example: Lorem Ipsum']
[main.v:16] dummy_methods['foo']: FunctionData{
    name: 'foo'
    attrs: ['example: Lorem Ipsum']
    args: []
    return_type: 1
    typ: 0
}
```

**Reply by Apocryphon-X** [reply]
My `reflection` module can be found here: https://gist.github.com/Apocryphon-X/6a6c535fd968f623037b42c6c2505056

**Reply by Apocryphon-X** [reply]
In the case of `vweb`, it seems like the function [`run_at<T>`](https://github.com/vlang/v/blob/f75d7eb11496d6c2cec2a5d3eca1feffebabc6f9/vlib/vweb/vweb.v#L386) uses reflection to parse the custom attributes from the `global_app` methods. That's why it needs an instance of the app as an argument.

**Reply by Apocryphon-X** [reply]
I fixed a little error in the module. Now it should work properly 👍🏻
